### DISCUSSION TITLE
Stepper motor support

### DISCUSSION THREAD

**Discussion by anjanbabu** [mwood77/winderoo#61] [Ideas]
This is an amazing project! love that it has a phone UI and OLED support! 
I see a lot of DIY watch winders use small stepper motors like the [28BYJ-48 unipolar stepper motor with the ULN2003 motor driver](https://randomnerdtutorials.com/esp32-stepper-motor-28byj-48-uln2003/), it would be nice if you could enable the Winderoo code to work with these motors as well since they're so cheap and accessible. These fancy 3D printable gyro winders also use the same motors. 
I did attempt to modify the code as [suggested](https://esp32.com/viewtopic.php?p=135879&sid=748c5ba2d0c311f7eed2870c0c47887c#p135879) by an ESP32 forum user, but I'm an ultra noob and I could not wrap my head around switching the PWM motor control code to a stepper motor one.

**Comment by mwood77** [comment]
Hi @anjanbabu - I didn't get notified about your topic (d'oh!).


> I did attempt to modify the code as [suggested](https://esp32.com/viewtopic.php?p=135879&sid=748c5ba2d0c311f7eed2870c0c47887c#p135879) by an ESP32 forum user, but I'm an ultra noob and I could not wrap my head around switching the PWM motor control code to a stepper motor one.

Ok, but you successfully got the code to compile without any changes? If so, then we can sort this out. In the root level of the project (not buried in any sub-folder) you'll find a file called `platform.ini`. 

That file looks like this:
```ini
; PlatformIO Project Configuration File
;
;   Build options: build flags, source filter
;   Upload options: custom upload port, speed and extra flags
;   Library options: dependencies, extra library storages
;   Advanced options: extra scripting
;
; Please visit documentation for the other options and examples
; https://docs.platformio.org/page/projectconf.html

[platformio]
description = IoT controlled smart watch winder

[env:esp32doit-devkit-v1]
platform = espressif32@^5.2.0
board = esp32doit-devkit-v1
framework = arduino
upload_speed = 115200
monitor_speed = 115200
build_src_filter = +<*> -<./angular/>
board_build.filesystem = littlefs
check_tool = cppcheck, clangtidy
build_flags = 
	-D OLED_ENABLED=false
	-D PWM_MOTOR_CONTROL=false
check_flags = 
	clangtidy: -fix-errors,--format-style=google
lib_deps = 
	esphome/AsyncTCP-esphome@^1.2.2
	esphome/ESPAsyncWebServer-esphome@^2.1.0
	Wire
	https://github.com/tzapu/WiFiManager.git
	https://github.com/bblanchon/ArduinoJson.git
	fbiego/ESP32Time@^2.0.0
	adafruit/Adafruit SSD1306@^2.5.9
	electromagus/ESPMX1508@^1.0.5
```

There's a line in this file:
```ini
	-D PWM_MOTOR_CONTROL=false
```

Change the value `false` to `true`, then rebuild the project. PWM control will now be enabled. If you have the stepper motor controller stuff on-hand, then read my next reply.

**Comment by mwood77** [comment]
As for converting the `MotorControl` class to stepper control, that's going to be a bit harder.

You'd need to modify this line to reference your 4 pins for the stepper motor controller board: https://github.com/mwood77/winderoo/blob/b2d59a803a1c96bf578acefc24e2de71b626fc6e/src/platformio/osww-server/src/main.cpp#L94

The result would be something like:
```cpp
# main.cpp

int pinA = 25;
int pinB = 26;
int pinC = 27;
int pinD = 28;

...
...
...

#if PWM_MOTOR_CONTROL   # < -- we're relying on you setting the build flag I talked about in the first reply, above!
	MotorControl motor(pinA, pinB, pinC, pinD);
#else
	MotorControl motor(directionalPinA, directionalPinB);
#endif

...
...
```

And in the `MotorControl.cpp` it'll be this-ish, and you'll need to match the class' interface in `MotorControl.h` too.

```cpp
# src/platformio/osww-server/src/utils/MotorControl.cpp


#include "MotorControl.h"
#include <Arduino.h>
#include <Stepper.h>
#include <freertos/FreeRTOS.h>
#include <freertos/task.h>

#define STEPS_PER_REVOLUTION 2048 // Number of steps per revolution for 28BYJ-48

// Global variables for task control
int8_t _motorDirection = 2;  // 0 for counterclockwise, 1 for clockwise, 2 to stop
int numSteps = 100;         // Number of steps to move (can be adjusted)
int stepSpeed = 10;         // Motor speed in RPM

MotorControl::MotorControl(int pinA, int pinB, int pinC, int pinD) 
    : _stepper(STEPS_PER_REVOLUTION, pinA, pinB, pinC, pinD)  // Initialize Stepper object
{
    // Set default motor speed
    _stepper.setSpeed(stepSpeed);
}

void MotorControl::clockwise()
{
    _motorDirection = 0;  // Set global direction to clockwise
    Serial.println("[STATUS] - Motor turning clockwise");
}

void MotorControl::countClockwise()
{
    _motorDirection = -1;  // Set global direction to counterclockwise
    Serial.println("[STATUS] - Motor turning counterclockwise");
}

void MotorControl::stop()
{
    _motorDirection = 2;  // Set global direction to stop
    Serial.println("[STATUS] - Motor stopped");
}

// FreeRTOS task to handle motor movement
void motorTask(void *parameter)
{
    MotorControl *motor = (MotorControl *)parameter;

    while (true)
    {
        if (_motorDirection != 2)
        {
            // Move motor in the specified direction
            motor->moveSteps(_motorDirection * numSteps);
        }
        else
        {
            // If _motorDirection is 0, stop the motor
            motor->stop();
            vTaskDelay(100 / portTICK_PERIOD_MS); // Check again after 100ms
        }
    }
}

void MotorControl::determineMotorDirectionAndBegin()
{
    // Create the FreeRTOS task for motor control
    xTaskCreate(motorTask, "Motor Task", 2048, this, 1, NULL);
}

void MotorControl::moveSteps(int steps)
{
    _stepper.step(steps);  // Move the motor by the specified number of steps
}

int MotorControl::getMotorDirection()
{
    return _motorDirection;
}

void MotorControl::setMotorDirection(int direction)
{
    _motorDirection = direction;
}

void MotorControl::setSpeed(int speed)
{
    stepSpeed = speed;
    _stepper.setSpeed(speed);
}
```

<hr>

I don't have any of the equipment on hand to test, but I think that's pretty much it.

**Comment by NickNerfLips** [comment]
Thank you for the guidance via the Reddit post, but I’m having issues with getting this to work.  I also tried the ChatGPT request to create the header file, but still getting errors.  I do appreciate all of your help thus far.

**Comment by NickNerfLips** [comment]
Chat GPT gives this:

#ifndef MOTORCONTROL_H

#define MOTORCONTROL_H

#include <Arduino.h>
#include <Stepper.h>

// Define the number of steps per revolution for 28BYJ-48
#define STEPS_PER_REVOLUTION 2048

class MotorControl {
public:
    // Constructor
    MotorControl(int pinA, int pinB, int pinC, int pinD);

    // Control methods
    void clockwise();
    void countClockwise();
    void stop();
    void determineMotorDirectionAndBegin();
    void setMotorDirection(int direction);
    void setSpeed(int speed);
    
    // Getter method
    int getMotorDirection();

private:
    // Method to move steps
    void moveSteps(int steps);

    // Stepper object
    Stepper _stepper;

    // Member variables
    static int _motorDirection;  // 0 for counterclockwise, 1 for clockwise, 2 to stop
    static int numSteps;         // Number of steps to move
    static int stepSpeed;        // Motor speed in RPM
};

#endif // MOTORCONTROL_H

**Reply by mwood77** [reply]
great - let me look into this.

I see you've ordered a SilentStepStick board & stepper motor. Can you provide links to the exact products you ordered? I assume the stepper motor chip is a TMC2XXX controller, but I need to make sure.

Looking closer at it, I think I'll need to spend some time figuring out how to best approach implementing this, while keeping it flexible for DC motors, or DC Motors w/ PWM. I'll likely also need to order some hardware too.

**Reply by NickNerfLips** [reply]
Absolutely.

Yes, the SilentStepStick driver board I'm using leverages a TMC2209:  https://www.amazon.com/dp/B0CC656Z59
The stepper motor I'm using is the STEPPERONLINE 17HE08-1004S:  https://www.amazon.com/dp/B0B93PNYCP
I added on a SilentStepStick Protector as well (prevent back current damage when unpowered):  https://www.aliexpress.us/item/2251832694497009.html

I'm using the ESP32 you recommended from AliExpress, although the USB-C variant.

I like this solution since it takes away the geartrain complexity as well as enables it to be near perfectly silent when in operation.

If I can get this up & running, I hope to reciprocate your assistance.  I'm a EE, but enjoy woodworking on the side.  I plan to make a watch winder chest variant out of wood as soon as I can get the software portion figured out first lol.

You definitely have the most complete & advanced open source watch winder plan out there, but likely also the most complex!
I really appreciate the efforts you put forth.  

Thanks and let me know if there's anything else I can help with.

**Reply by mwood77** [reply]
Alright I've ordered the same hardware, but it'll take a week or two to arrive. I can take some pot-shots at the code in the meantime though.

Are you familiar with how git works? If I create a branch with changes, would you be willing to try them out?

> You definitely have the most complete & advanced open source watch winder plan out there, but likely also the most complex!
I really appreciate the efforts you put forth.

Thanks! It's slowly getting more and more complex - the projects originally built on one another, but I'm starting to wonder if they should be totally separate. However, Winderoo is also meant to be a drop-in logic controller for existing consumer watch winders, sooo...

**Reply by mwood77** [reply]
OK, here's a build which _might work._ It compiles OK, and appears to be functional on my demo boards. But I cannot confirm the TMC2209 and stepper motor are handled correctly, as I don't have the hardware on hand.

I've had to spec the following pins for your SilentStepStick driver board:
```cpp
    const int RX_PIN = 5;
    const int TX_PIN = 26;
    const uint8_t HARDWARE_ENABLE_PIN = 4;
```

I've also guessed at the motor speed. If the motor actually spins, can you please confirm that it takes approximately 8 seconds to complete a full revolution?

Let me know if you need help pulling this, but here's the code: https://github.com/mwood77/winderoo/pull/69
